### PR TITLE
Create devnets

### PR DESCRIPTION
**Pull request by felipemadero**
Closes https://github.com/ava-labs/avalanche-cli/issues/1150 https://github.com/ava-labs/avalanche-cli/issues/1111 

Step 1:
- create node to create a devnet if --devnet flag is set
- creating genesis for devnet and avalanchego configs for devnet
- encapsulate network id and network endpoint under models.Network
- add models.ClusterConfig to get both nodes and network def of cluster
- rename models.ClusterConfig to models.ClustersConfig

https://asciinema.org/a/619645

### CODE REVIEW

File: cmd/keycmd/list.go [hunk 1]
```diff
@@ -119,18 +119,13 @@ func getClients(networks []models.Network, cchain bool) (
...
-	}
 	var err error
 	pClients := map[models.Network]platformvm.Client{}
 	cClients := map[models.Network]ethclient.Client{}
 	for _, network := range networks {
-		pClients[network] = platformvm.NewClient(apiEndpoints[network])
+		pClients[network] = platformvm.NewClient(network.Endpoint)
 		if cchain {
-			cClients[network], err = ethclient.Dial(fmt.Sprintf("%s/ext/bc/%s/rpc", apiEndpoints[network], "C"))
+			cClients[network], err = ethclient.Dial(fmt.Sprintf("%s/ext/bc/%s/rpc", network.Endpoint, "C"))
```
Comment (arturrez) [inline cmd/keycmd/list.go:128]: nit: provide func to return `fmt.Sprintf("%s/ext/bc/%s/rpc", network.Endpoint, "C")` as part of the model Network

File: cmd/keycmd/list.go [hunk 2]
```diff
@@ -343,7 +330,7 @@ func getPChainAddrInfo(
 		chain:   "P-Chain (Bech32 format)",
 		address: pChainAddr,
 		balance: balance,
-		network: network.String(),
+		network: network.Kind.String(),
```
Comment (arturrez) [inline cmd/keycmd/list.go:333]: nit: provide func for this

File: cmd/keycmd/transfer.go [hunk 3]
```diff
@@ -155,15 +154,15 @@ func transferF(*cobra.Command, []string) error {
 
 	var network models.Network
 	if local {
-		network = models.Local
+		network = models.LocalNetwork
 	}
 	if testnet {
```
Comment (arturrez) [inline cmd/keycmd/transfer.go:159]: nit: else if

File: cmd/keycmd/transfer.go [hunk 4]
```diff
@@ -155,15 +154,15 @@ func transferF(*cobra.Command, []string) error {
...
-		network = models.Fuji
+		network = models.FujiNetwork
 	}
 	if mainnet {
-		network = models.Mainnet
+		network = models.MainnetNetwork
 	}
-	if network == models.Undefined {
+	if network.Kind == models.Undefined {
```
Comment (arturrez) [inline cmd/keycmd/transfer.go:165]: nit: just `else` after bunch of `else if`

File: cmd/nodecmd/create.go [hunk 5]
```diff
@@ -207,12 +226,20 @@ func createNodes(_ *cobra.Command, args []string) error {
 		return err
 	}
 	createdAnsibleHostIDs := strings.Join(ansibleHostIDs, ",")
-	if err = runAnsible(inventoryPath, avalancheGoVersion, clusterName, createdAnsibleHostIDs); err != nil {
+	if err = runAnsible(network, inventoryPath, avalancheGoVersion, clusterName, createdAnsibleHostIDs); err != nil {
```
Comment (arturrez) [inline cmd/nodecmd/create.go:229]: nit: network shold be second argument as first one is always `inventoryPath`

File: cmd/nodecmd/create.go [hunk 6]
```diff
@@ -298,14 +332,21 @@ func setupAnsible(clusterName string) error {
 	return updateAnsiblePublicIPs(clusterName)
 }
 
-func runAnsible(inventoryPath, avalancheGoVersion, clusterName, ansibleHostIDs string) error {
+func runAnsible(network models.Network, inventoryPath, avalancheGoVersion, clusterName, ansibleHostIDs string) error {
```
Comment (arturrez) [inline cmd/nodecmd/create.go:335]: nit: network should be second after inventoryPath

File: cmd/nodecmd/devnet.go [hunk 7]
```diff
@@ -0,0 +1,211 @@
...
+	"github.com/ava-labs/avalanche-cli/pkg/utils"
+	"github.com/ava-labs/avalanche-cli/pkg/ux"
+	"github.com/ava-labs/avalanchego/config"
+	coreth_params "github.com/ava-labs/coreth/params"
+)
+
+// difference between unlock schedule locktime and startime in original genesis
+const (
+	genesisLocktimeStartimeDelta = 2836800
+	hexa0Str                     = "0x0"
```
Comment (arturrez) [inline cmd/nodecmd/devnet.go:27]: can we use hex.EncodeString to all hex values to make it more readable?

File: cmd/nodecmd/devnet.go [hunk 8]
```diff
@@ -0,0 +1,211 @@
...
+	cChainGenesisMap["config"] = coreth_params.AvalancheLocalChainConfig
+	cChainGenesisMap["nonce"] = hexa0Str
+	cChainGenesisMap["timestamp"] = hexa0Str
+	cChainGenesisMap["extraData"] = "0x00"
+	cChainGenesisMap["gasLimit"] = "0x5f5e100"
+	cChainGenesisMap["difficulty"] = hexa0Str
+	cChainGenesisMap["mixHash"] = "0x0000000000000000000000000000000000000000000000000000000000000000"
+	cChainGenesisMap["coinbase"] = "0x0000000000000000000000000000000000000000"
+	cChainGenesisMap["alloc"] = map[string]interface{}{
+		"8db97C7cEcE249c2b98bDC0226Cc4C2A57BF52FC": map[string]interface{}{
```
Comment (arturrez) [inline cmd/nodecmd/devnet.go:41]: what is this address?

File: cmd/nodecmd/devnet.go [hunk 9]
```diff
@@ -0,0 +1,211 @@
...
+			"balance": "0x295BE96E64066972000000",
```
Comment (arturrez) [inline cmd/nodecmd/devnet.go:42]: hex.EncodeString

File: cmd/nodecmd/devnet.go [hunk 10]
```diff
@@ -0,0 +1,211 @@
...
+		initialStakers = append(initialStakers, initialStaker)
+	}
+	genesisMap["initialStakeDuration"] = 31536000
+	genesisMap["initialStakeDurationOffset"] = 5400
+	genesisMap["initialStakers"] = initialStakers
+	lockTime := startTime + genesisLocktimeStartimeDelta
+	allocations := []interface{}{}
+	alloc := map[string]interface{}{
+		"avaxAddr":      walletAddr,
+		"ethAddr":       "0xb3d82b1367d362de99ab59a658165aff520cbd4d",
```
Comment (arturrez) [inline cmd/nodecmd/devnet.go:81]: should be in const

File: cmd/nodecmd/devnet.go [hunk 11]
```diff
@@ -0,0 +1,211 @@
...
+		"unlockSchedule": []interface{}{
+			map[string]interface{}{"amount": 10000000000000000, "locktime": lockTime},
+		},
+	}
+	allocations = append(allocations, alloc)
+	genesisMap["allocations"] = allocations
+	genesisMap["initialStakedFunds"] = []interface{}{
+		stakingAddr,
+	}
+	genesisMap["message"] = "{{ fun_quote }}"
```
Comment (arturrez) [inline cmd/nodecmd/devnet.go:102]: `fun_quote` ?

File: cmd/nodecmd/devnet.go [hunk 12]
```diff
@@ -0,0 +1,211 @@
...
+		n, err := getNodeID(app.GetNodeInstanceDirPath(s))
+		return n.String(), err
+	})
+	if err != nil {
+		return err
+	}
+
+	// set devnet network
+	network := models.DevnetNetwork
+	network.Endpoint = "http://" + ansibleHosts[ansibleHostIDs[0]].IP + ":9650"
```
Comment (arturrez) [inline cmd/nodecmd/devnet.go:136]: provide func in models that returns network struct with input IP address

File: cmd/subnetcmd/helpers.go [hunk 13]
```diff
@@ -0,0 +1,91 @@
...
+		network = models.LocalNetwork
+	case useDevnet:
+		network = models.DevnetNetwork
+	case useFuji:
+		network = models.FujiNetwork
+	case useMainnet:
+		network = models.MainnetNetwork
+	}
+
+	if endpoint != "" {
```
Comment (arturrez) [inline cmd/subnetcmd/helpers.go:47]: no need to pass endpoint if you can implement comment above and func that can return network struct with endpoint set from IP

File: cmd/subnetcmd/join.go [hunk 14]
```diff
@@ -704,7 +683,18 @@ func promptStakeAmount(subnetName string, isValidator bool, network models.Netwo
 	return initialSupply, nil
 }
 
-func printJoinCmd(subnetID string, networkID string, vmPath string) {
+func printJoinCmd(subnetID string, network models.Network, vmPath string) {
+	networkIDValue := ""
+	switch network.Kind {
```
Comment (arturrez) [inline cmd/subnetcmd/join.go:688]: this should be part of the network struct like network.Tag() or something like that

Comment (arturrez) [reply cmd/subnetcmd/join.go:688]: so it would look like `ntagetworkIDValue := network.Tag()`

Comment (arturrez) [reply cmd/subnetcmd/join.go:688]: maybe better name

File: pkg/plugins/avagoConfig.go [hunk 15]
```diff
@@ -104,7 +105,19 @@ func EditConfigFile(
 	// Banf.10 changes from "whitelisted-subnets" to "track-subnets"
 	delete(avagoConfig, "whitelisted-subnets")
 	avagoConfig["track-subnets"] = newVal
-	avagoConfig["network-id"] = networkID
+
+	networkIDValue := ""
+	switch network.Kind {
```
Comment (arturrez) [inline pkg/plugins/avagoConfig.go:110]: should be separate func for network struct see above

File: cmd/nodecmd/devnet.go (hunk 8 above)
Comment (felipemadero) [reply cmd/nodecmd/devnet.go:41]: added constant for it

File: cmd/nodecmd/devnet.go (hunk 11 above)
Comment (felipemadero) [reply cmd/nodecmd/devnet.go:102]: just copying local genesis from avalanchego. it is fun?

File: cmd/subnetcmd/helpers.go [hunk 16]
```diff
@@ -0,0 +1,91 @@
...
+		endpoint, err := app.Prompt.CaptureString("Devnet Network Endpoint")
+		if err != nil {
+			return err
+		}
+		network.Endpoint = endpoint
+	}
+	return nil
+}
+
+func GetNetworkFromCmdLineFlags(
```
[the remark this answers is not in the document]
Comment (felipemadero) [reply cmd/subnetcmd/helpers.go:26]: no, this is a function to process cmdline flags it needs to receive all the flags

File: cmd/subnetcmd/helpers.go (hunk 13 above)
Comment (felipemadero) [reply cmd/subnetcmd/helpers.go:47]: the endpoint is when passed as cmdline flag, it also enables the user to pass the port in flag

Comment (felipemadero) [reply cmd/subnetcmd/helpers.go:47]: this function returns a network from cmdline flags

File: cmd/keycmd/transfer.go [hunk 17]
```diff
@@ -154,17 +153,14 @@ func transferF(*cobra.Command, []string) error {
...
+	case local:
+		network = models.LocalNetwork
+	case testnet:
+		network = models.FujiNetwork
+	case mainnet:
+		network = models.MainnetNetwork
+	default:
 		networkStr, err := app.Prompt.CaptureList(
 			"Network to use",
 			[]string{models.Mainnet.String(), models.Fuji.String(), models.Local.String()},
```
Comment (arturrez) [inline cmd/keycmd/transfer.go:166]: should be Name()?

File: cmd/primarycmd/addValidator.go [hunk 18]
```diff
@@ -122,12 +122,12 @@ func addValidator(_ *cobra.Command, _ []string) error {
...
 	case validateMainnet:
-		network = models.Mainnet
+		network = models.MainnetNetwork
 	}
 
-	if network == models.Undefined {
+	if network.Kind == models.Undefined {
 		networkStr, err := app.Prompt.CaptureList(
 			"Choose a network to add validator to.",
 			[]string{models.Fuji.String(), models.Mainnet.String()},
```
Comment (arturrez) [inline cmd/primarycmd/addValidator.go:133]: same

File: cmd/subnetcmd/addDelegator.go [hunk 19]
```diff
@@ -104,7 +96,7 @@ func addPermissionlessDelegator(_ *cobra.Command, args []string) error {
 	if useLedger && keyName != "" {
 		return ErrMutuallyExlusiveKeyLedger
 	}
-	subnetID := sc.Networks[network.String()].SubnetID
+	subnetID := sc.Networks[network.Name()].SubnetID
 	if os.Getenv(constants.SimulatePublicNetwork) != "" {
 		subnetID = sc.Networks[models.Local.String()].SubnetID
```
Comment (arturrez) [inline cmd/subnetcmd/addDelegator.go:101]: Name()

**Review by arturrez — APPROVED** [review]

File: cmd/keycmd/transfer.go (hunk 17 above)
Comment (felipemadero) [reply cmd/keycmd/transfer.go:166]: Nop

**Pull request merged**